src/OrbitQt/AnnotatingSourceCodeHandler.h

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_QT_ANNOTATING_SOURCE_CODE_HANDLER_H_
+#define ORBIT_QT_ANNOTATING_SOURCE_CODE_HANDLER_H_
+
+#include <absl/types/span.h>
+
+#include <QObject>
+#include <QPointer>
+#include <chrono>
+#include <functional>
+
+#include "App.h"
+#include "CodeReport/AnnotatingLine.h"
+#include "CodeReport/DisassemblyReport.h"
+#include "CodeViewer/OwningDialog.h"
+#include "ObjectUtils/ElfFile.h"
+#include "OrbitBase/Future.h"
+#include "OrbitBase/Result.h"
+#include "QtUtils/MainThreadExecutorImpl.h"
+#include "capture_data.pb.h"
+#include "symbol.pb.h"
+
+namespace orbit_qt {
+
+// AnnotatingSourceCodeHandler manages the retrieval of debug information, determining the source
+// code location, and matching source code lines to assembly instructions.
+//
+// It is meant to be used with orbit_code_viewer::OwningDialog. It uses the status message option
+// of orbit_code_viewer::Dialog to interact with the user when manual input is needed.
+//
+// The typical way of using that class is by allocating an instance on the heap and make the dialog
+// its parent:
+// new AnnotatingSourceCodeHandler(dialog, function_info, report, app, dialog);
+//
+// Qt will manage its lifetime and delete it when the dialog is also deleted.
+class AnnotatingSourceCodeHandler : public QObject {
+  Q_OBJECT
+
+ public:
+  // Same interface as OrbitApp::RetrieveModuleWithDebugInfo;
+  using RetrieveModuleWithDebugInfoCallback =
+      std::function<orbit_base::Future<ErrorMessageOr<std::filesystem::path>>(const std::string&,
+                                                                              const std::string&)>;
+  explicit AnnotatingSourceCodeHandler(const QPointer<orbit_code_viewer::OwningDialog>& dialog,
+                                       orbit_client_protos::FunctionInfo function_info,
+                                       orbit_code_report::DisassemblyReport report,
+                                       RetrieveModuleWithDebugInfoCallback callback,
+                                       QObject* parent = nullptr);
```

Comment:
`QObject* parent = nullptr` I assume you want to adhere to qt style, but iff a `orbit_code_viewer::OwningDialog` is required anyways, maybe skip that additional parent parameter and always use `dialog` as parent?

I bet you thought about it and have a reason why the current way is good. I'd be interested in that thought

Reply:
I was thinking exactly the same when reading this. What is the logical difference between owning dialog and parent, especially if the owning dialog is supposed to manage the lifetime?

Also about argument order, should (if it really is required separately) `parent` be the first or second argument, following how QT is usually defining the constructors?

Reply:
Also, can you elaborate why `QPointer` instead of `OwningDialog*`? AFAIK we're not using this anywhere else?

Reply:
This is implemented like this for a couple of reasons:

1. `OwningDialog` doesn't need to manage the lifetime. For example the test creates the object on the stack.
2. I want lifetime management to be explicit here. The user of this class needs to know what they are doing. If I remove the `parent` parameter it gets implicit and it gets unclear to users what is happening when they just read the line that is instantiating this object. By sticking to the Qt-standard-pattern. This is much more clear.
3. We need a QPointer to know when `OwningDialog` gets out of scope. A raw pointer might get dangling. That's why you see a couple of `if (dialog_.IsNull()) return;` statements in the code. Note, that this requirement won't go away even if we made the `parent` implicit and always set it to `dialog_` because all children are get destroyed after `OwningDialog` gets destroyed.

We use QPointer at a couple of locations in `OrbitGgp` and `OrbitSshQt` and `QtUtils`.

Reply:
I also don't believe that "not used anywhere else" is a strong argument for not using it.

Reply:
Not a strong argument, but I do think "consistency" is an argument. But sure, if it makes sense then this should not be a reason against it.

Regarding lifetime, isn't the Qt standard pattern that parent manages the lifetime, or am I mixing this up now?

Reply:
```
Not a strong argument, but I do think "consistency" is an argument. But sure, if it makes sense then this should not be a reason against it.
```

It's not about consistency. QPointer solves a problem which can't be solved by any other mechanism.

```
Regarding lifetime, isn't the Qt standard pattern that parent manages the lifetime, or am I mixing this up now?
```
No you are correct. Parents delete children when they get deleted theirselves. But this happens in the destructor of QObject, so after `OwningDialog`'s destructor has been executed. That means there is a small period of time when `AnnotatingSourceCodeHandler` is alive and `OwningDialog` is not.

This is not a problem as long as `AnnotatingSourceCodeHandler` does not access `OwningDialog` during that time. And we ensure that by checking if the QPointer is null. Qt's data views and data models do similar things to ensure that destruction oder doesn't matter.

If we want to avoid the whole QPointer-thing we can make `OwningDialog` a member of `AnnotatingSourceCodeHandler`. That would ensure a strict lifetime hierarchy and the problem can't occur. Maybe that's not the worst idea. WDYT? Of course we would rename it to `AnnotatingSourceCodeDialog` or something similar.

Reply:
It's not so much about `QPointer`, I've just not seen us use that before, that's why it stood out to me, and really the thing that struck me most odd was that we pass `dialog` and `parent` separately. Getting rid of this would feel clearer and more "familiar" to me.

But I don't have a super strong opinion now that I understand the reasoning behind it, I'm also fine with leaving it as is, maybe then this reasoning should be pointed out in a comment.

Reply:
Okay, I will have a quick look if changing this makes it simpler.

Reply:
A refactored the class and there is now an `AnnotatingSourceCodeDialog` which inherits from `orbit_code_viewer::Dialog`. PTAL

Reply:
I like this, thanks!